moabb/pipelines/classification.py

```diff
@@ -511,3 +534,146 @@ def predict_proba(self, X):
+     to compute the reference signal used in CCA [1]_.
+
+    Parameters
+    ----------
+    freqs : dict with n_classes keys
+        Frequencies corresponding to the SSVEP stimulation frequencies.
+        They are used to identify SSVEP classes presents in the data.
+
+    n_filters: int
+        Number of multisets spatial filters used per sample data.
+        It corresponds to the number of eigen vectors taken the solution of the
+        MAXVAR objective function as formulated in Eq.5 in [1]_.
+
+    n_components: int
+        Number of CCA components to be used.
+
+
+    References
+    ----------
+
+    .. [1] Zhang, Y.U., Zhou, G., Jin, J., Wang, X. and Cichocki, A. (2014). Frequency
+           recognition in SSVEP-based BCI using multiset canonical correlation analysis.
+           International journal of neural systems, 24(04), p.1450013.
+           https://doi.org/10.1142/S0129065714500130
+    """
+
+    def __init__(self, freqs, n_filters=1, n_components=1, n_jobs=1):
+        self.n_jobs = n_jobs
+        self.n_filters = n_filters
+        self.freqs = freqs
+        self.n_components = n_components
+        self.cca = CCA(n_components=n_components)
+
+    def fit(self, X, y, sample_weight=None):
+        """
+        Compute the optimized reference signal at each stimulus frequency
+        """
+        self.classes_ = np.unique(y)
+        self.one_hot = {}
+        for i, k in enumerate(self.classes_):
+            self.one_hot[k] = i
+        n_trials, n_channels, n_times = X.shape
+
+        # Whiten signal in parallel
+        if self.n_jobs == 1:
+            X_white = [_whitening(X_i) for X_i in X]
+        else:
+            X_white = Parallel(n_jobs=self.n_jobs)(delayed(_whitening)(X_i) for X_i in X)
+        X_white = np.stack(X_white, axis=0)
+
+        Y = X_white.transpose(2, 0, 1).reshape(-1, n_times)
+        # R = np.cov(Y)
+        # or more similar to the article:
+        R = Y @ Y.T
+        # S = np.diag(np.diag(R)) # This does not match the definition in the article
+        # S exactly as defined in the article
+        mask = np.kron(
+            np.eye(n_trials), np.ones((n_channels, n_channels))
+        )  # block diagonal mask
+        S = R * mask
+
+        # Get W
+        _, tempW = linalg.eigh(
+            R - S, S, subset_by_index=[R.shape[1] - self.n_filters, R.shape[1] - 1]
+        )
+        W = np.reshape(tempW, (n_trials, n_channels, self.n_filters))
+
+        # Normalise W
+        W = W / np.linalg.norm(W, axis=0, keepdims=True)
+        print(f"[DEBUG] W.shape: {W.shape}")
+
+        # get Z in parallel
+        if self.n_jobs == 1:
+            Z = [np.dot(W_i.T, X_white_i) for W_i, X_white_i in zip(W, X_white)]
+        else:
+            Z = Parallel(n_jobs=self.n_jobs)(
+                delayed(np.dot)(W_i.T, X_white_i) for W_i, X_white_i in zip(W, X_white)
+            )
+        Z = np.stack(Z, axis=0)
```

Comment:
Use NumPy matrix multiplication broadcasting:
```suggestion
        Z = W.transpose((0, 2, 1)) @ X_white
```

Reply:
Thanks @qbarthelemy!